dev/src/transaction.ts

```diff
@@ -552,3 +561,14 @@ function validateReadOptions(
     }
   }
 }
+
+function isRetryableTransactionError(error: Error): boolean {
+  if (error instanceof GoogleError || 'code' in error) {
+    // In transactions, the backend returns code ABORTED for reads that fail
+    // with contention. These errors should be retried for both GoogleError
+    // and GoogleError-alike errors (in case the prototype hierarchy gets
```

Comment:
Could this check for `error.name === 'GoogleError'` instead?

When implementing gRPC servers, it's not unthinkable that an ABORTED error is thrown from inside the transaction, with the expectation that it's returned to the client.

Reply:
This should accept any object that has a "code" property. In the environment you are describing, will this property be stripped?

Reply:
Let me rephrase.

There are many ways of wrongly implementing a transaction function, to do with side-effects, or forgetting to use the transaction itself to perform queries.

This new `ABORTED` handling adds more ways to get it wrong.

There is an assumption that all `GoogleError` instances came from transaction methods. That may not be the case if I use another GCP SDK within my function. I know, I shouldn't be, but somebody still might.

It also won't be the case if I call a Firestore method on a `DocumentReference` which then throws an `ABORTED` error.

These assumptions are then generalized by testing if the thrown error has a `.code = 10` property. I have a `ServiceError` object that is thrown with a gRPC response code. If for some reason I throw that error inside a transaction function, my transaction may be retried even if it succeeded. And again, sure, I suppose I shouldn't be throwing that error, but somebody still might.

A safer approach may be to register errors, thrown as part of the actual transaction, in a `WeakMap`. Then as they propagate through user code, when you catch them again you can check whether they're genuine transaction errors, and then check their code.

Reply:
A couple things to note:
- A read outside of a transaction should never result in Code ABORTED. This error should only happen during a transactional read.
- If a user throws an error with `code: 10`, we will retry the transaction. In general, transaction functions should be written so they can be re-run, and re-running them more often than needed should not cause any harm.

I understand that we control the errors that we are catching here. We could tighten this check to make sure that the error is indeed a GoogleError, but for now I decided to err on the side of caution and also retry GoogleError-alike errors. This allows us to catch errors with code properties that were created by our dependencies before we converted everything to use the proper GoogleError type.